FlowCrypt/Controllers/Attachment/AttachmentViewController.swift

```diff
@@ -0,0 +1,124 @@
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+
+import UIKit
+import WebKit
+import FlowCryptUI
+import Combine
+
+@MainActor
+final class AttachmentViewController: UIViewController {
+
+    private lazy var webView = WKWebView()
+    private let file: FileType
+    private let shouldShowDownloadButton: Bool
+
+    private let filesManager: FilesManagerType
+    private lazy var attachmentManager = AttachmentManager(
+        controller: self,
+        filesManager: filesManager
+    )
+
+    private let errorLabel: UILabel = {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.textAlignment = .center
+        label.text = "no_preview_avalable".localized
+        label.isHidden = true
+        label.textColor = .textColor
+        label.font = .systemFont(ofSize: 12)
+        return label
+    }()
+
+    init(
+        file: FileType,
+        shouldShowDownloadButton: Bool,
+        filesManager: FilesManagerType = FilesManager()
+    ) {
+        self.file = file
+        self.shouldShowDownloadButton = shouldShowDownloadButton
+        self.filesManager = filesManager
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    @available(*, unavailable)
+    required init?(coder _: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        setupNavigationBar()
+        addWebView()
+        showSpinner()
+        title = file.name
+        load()
+    }
+
+    private func setupNavigationBar() {
+        guard shouldShowDownloadButton else { return }
+        navigationItem.rightBarButtonItem = NavigationBarItemsView(
+            with: [
+                NavigationBarItemsView.Input(
+                    image: UIImage(named: "download")?.tinted(.gray),
+                    title: "save".localized,
+                    onTap: { [weak self] in
+                        guard let self = self else { return }
+                        self.attachmentManager.download(self.file)
+                    }
+                )
+            ]
+        )
+    }
+}
+
+extension AttachmentViewController: WKNavigationDelegate {
+
+    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
+        errorLabel.isHidden = false
+        hideSpinner()
+    }
+
+    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        hideSpinner()
+    }
+}
+
+private extension AttachmentViewController {
+
+    private func load() {
+        webView.load(
+            file.data,
+            mimeType: file.name.mimeType,
+            characterEncodingName: "UTF-8",
+            // It is a URL that should be used to resolve relative URLs within the document.
+            // As we don't resolve any relative URLs we pass just NSURL()
+            baseURL: NSURL() as URL
+        )
+    }
```

Comment:
Likely caused by the `characterEncodingName` here - it's treating it as text regardless of actual content.

That's why only file types that are inherently textual continue to work.

Reply:
I checked `.png` and `.jpg` - work well

Reply:
@sosnovsky Can we still use the original method that opens a file from url, instead of giving it raw bytes, but instead of file url that points to a local filestem, we would give it a memory-based blob url? Is there such a concept on iOS?

Reply:
Answering my own question: I think yes https://stackoverflow.com/a/23647182